openssl/src/x509/mod.rs

```diff
@@ -443,6 +460,153 @@ impl X509Ref {
         }
     }
 
+    /// Returns this certificate's key usage.
+    ///
+    /// This corresponds to [`X509_get_ext_d2i`] called with `NID_key_usage`.
+    ///
+    /// [`X509_get_ext_d2i`]: https://www.openssl.org/docs/man1.1.0/crypto/X509_get_ext_d2i.html
+    pub fn key_usage(&self) -> Option<KeyUsageFlags> {
+        unsafe {
+            let key_usage_ptr = ffi::X509_get_ext_d2i(
+                self.as_ptr(),
+                ffi::NID_key_usage,
+                ptr::null_mut(),
+                ptr::null_mut(),
+            );
+
+            if key_usage_ptr.is_null() {
+                return None;
+            }
+
+            let key_usage_str = Asn1BitString::from_ptr(key_usage_ptr as *mut _);
+            if key_usage_str.len() == 0 {
+                return None;
+            }
+
+            let mut flags: u32 = 0;
+            let num_bytes = cmp::min(key_usage_str.len(), mem::size_of::<u32>());
+            let data = key_usage_str.as_slice();
+
+            for i in 0..num_bytes {
+                flags ^= (data[i] as u32) << ((num_bytes - (i + 1)) * 8);
+            }
+
+            println!("{}", num_bytes);
+
+            KeyUsageFlags::from_bits(flags)
+        }
+    }
+
+    /// Returns this certificate's extended key usage, if it exists.
+    ///
+    /// This corresponds to [`X509_get_ext_d2i`] called with `NID_ext_key_usage`.
+    ///
+    /// [`X509_get_ext_d2i`]: https://www.openssl.org/docs/man1.1.0/crypto/X509_get_ext_d2i.html
+    pub fn ext_key_usage(&self) -> Option<Stack<Asn1Object>> {
+        unsafe {
+            let ext_key_usage = ffi::X509_get_ext_d2i(
+                self.as_ptr(),
+                ffi::NID_ext_key_usage,
+                ptr::null_mut(),
+                ptr::null_mut(),
+            );
+
+            if ext_key_usage.is_null() {
+                None
+            } else {
+                Some(Stack::from_ptr(ext_key_usage as *mut _))
+            }
+        }
+    }
+
+    /// Returns this certificate `AuthorityKeyId`, if it exists.
+    ///
+    /// This corresponds to [`X509_get_ext_d2i`] called with `NID_authority_key_identifier`.
+    ///
+    /// [`X509_get_ext_d2i`]: https://www.openssl.org/docs/man1.1.0/crypto/X509_get_ext_d2i.html
+    pub fn authority_keyid(&self) -> Option<AuthorityKeyId> {
```

Comment:
`key_id`.

Reply:
This naming convention was chosen to be more in-line with OpenSSL's use of `keyid` and [`AuthorityKeyIdentifier::keyid()`](https://docs.rs/openssl/0.10.20/openssl/x509/extension/struct.AuthorityKeyIdentifier.html#method.keyid), but I agree.